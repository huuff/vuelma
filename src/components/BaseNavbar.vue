<template>
<nav 
  class="navbar" 
  role="navigation" 
  aria-label="main navigation"
  @blur="showMobile = false"
  >
  <div class="navbar-brand">
    <classed-slot children-class="navbar-item">
      <slot name="brandItems"></slot>
    </classed-slot>
    <a 
      role="button" 
      class="navbar-burger"
      :class="{ 'is-active': showMobile }"
      aria-label="menu"
      aria-expanded="false"
      @click="toggleMobileMenu"
    >
      <span aria-hidden="true"></span>
      <span aria-hidden="true"></span>
      <span aria-hidden="true"></span>
    </a>
  </div>
  <div 
    class="navbar-menu"
    :class="{ 'is-active': showMobile }"
  >
  <div class="navbar-start" v-if="$slots.start">
    <classed-slot children-class="navbar-item">
      <slot name="start"></slot>
    </classed-slot>
  </div>
  <div class="navbar-end" v-if="$slots.end">
    <classed-slot children-class="navbar-item">
      <slot name="end"></slot>
    </classed-slot>
  </div>
  </div>
</nav>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import ClassedSlot from '@/components/slots/ClassedSlot';

const showMobile = ref(false);

function toggleMobileMenu() {
  showMobile.value = !showMobile.value;
  if (showMobile.value) {
    window.requestAnimationFrame(() => {
      document.addEventListener('click', () => {
          showMobile.value = false;
      }, { once: true})
    });
  }
}
</script>
