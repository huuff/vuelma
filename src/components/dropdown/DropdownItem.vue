<template>
<slot></slot>
</template>

<script setup lang="ts">
import { OptionalId } from '@/util/optional-id';

export interface DropdownItemProps extends OptionalId {
  text: string;
  itemId?: string;
}

const props = defineProps<DropdownItemProps>();

</script>
