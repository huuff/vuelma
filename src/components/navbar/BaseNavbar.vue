<template>
<nav 
  class="navbar" 
  role="navigation" 
  aria-label="main navigation"
  @blur="showMobile = false"
  >
  <div class="navbar-brand">
    <slot name="brandItems"></slot>
    <a 
      role="button" 
      class="navbar-burger"
      :class="{ 'is-active': showMobile }"
      aria-label="menu"
      aria-expanded="false"
      @click="toggleMobileMenu"
    >
      <span aria-hidden="true"></span>
      <span aria-hidden="true"></span>
      <span aria-hidden="true"></span>
    </a>
  </div>
  <div 
    class="navbar-menu"
    :class="{ 'is-active': showMobile }"
  >
  <div v-if="$slots.start"
    class="navbar-start"
  >
    <slot name="start"></slot>
  </div>
  <div v-if="$slots.end"
    class="navbar-end"
  >
    <slot name="end"></slot>
  </div>
  </div>
</nav>
</template>

<script setup lang="ts">
import { provide, ref, computed } from 'vue';

const props = defineProps<{
  active?: string
}>();

const emit = defineEmits<{
  (event: "update:active", title: string): void;
}>();

const showMobile = ref(false);
const internalActive = ref<string | undefined>(undefined);

// Gets the active from prop if it exists, or from an
// internal state if it exists. Updates both the prop
// and th interal state
const actualActive = computed({
  get: () => props.active ?? internalActive.value,
  set: (itemId: string | undefined) => {
    if (itemId) {
      internalActive.value = itemId;
      emit('update:active', itemId);
    }
  }
});

function toggleMobileMenu() {
  showMobile.value = !showMobile.value;
  if (showMobile.value) {
    requestAnimationFrame(() => {
      document.addEventListener('click', () => {
          console.log("Hiding on mobile");
          showMobile.value = false;
      }, { once: true})
    });
  }
}

provide('setActiveNavbarItem', (id: string) => actualActive.value = id);
provide('activeNavbarItem', () => actualActive.value);
</script>
