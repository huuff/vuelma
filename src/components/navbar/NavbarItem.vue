<template>
<slot></slot>
</template>

<script setup lang="ts">
import { Component } from 'vue';
import { OptionalId } from '@/util/optional-id';

export interface NavbarItemProps extends OptionalId {
  tag?: string | Component;
  titleText: string;
  itemId?: string;
}

const props = defineProps<NavbarItemProps>();

</script>
