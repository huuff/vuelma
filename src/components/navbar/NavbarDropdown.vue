<template>
<div class="navbar-item has-dropdown" :class="{
    'is-active': actualOpen,
    'has-dropdown-up': direction === 'up',
    'is-hoverable': hoverable,
    'is-right': isRight,
  }">
  <a class="navbar-link" @click="actualOpen = !actualOpen">
    {{ triggerText }}
  </a>
  <render></render>
</div>
</template>

<script setup lang="tsx">
// TODO: arrowless
import { useSlots, VNode, h, toRef } from 'vue';
import DropdownItem, { DropdownItemProps } from '@/components/dropdown/DropdownItem.vue'
import DropdownDivider from '../dropdown/DropdownDivider.vue';
import { DropdownDirection } from '@/components/dropdown/BaseDropdown.vue';
import classnames from 'classnames';
import { useOptionalTwoWayBinding } from '@/composables/optional-two-way-binding';
import { useCloseOnClickOutside } from '@/composables/close-on-click-outside';
import partial from "lodash/partial";

const props = withDefaults(defineProps<{
  triggerText?: string;
  activeItem?: string;
  direction?: DropdownDirection;
  open?: boolean;
  hoverable?: boolean;
  isRight?: boolean;
}>(), {
  open: undefined,
  direction: "down",
});

const emit = defineEmits<{
  (event: "update:activeItem", itemId?: string): void;
  (event: "update:open", open: boolean): void;
}>();

//eslint-disable-next-line
//@ts-ignore
const actualActiveItem = useOptionalTwoWayBinding(undefined, toRef(props, "activeItem"), partial(emit, "update:activeItem"));

//eslint-disable-next-line
//@ts-ignore
const actualOpen = useOptionalTwoWayBinding(false, toRef(props, "open"), partial(emit, "update:open"));
useCloseOnClickOutside(actualOpen);

function renderNode(node: VNode): VNode {
  if (node.type === DropdownItem) {
    // TODO: Almost the exact same code as the render node of BaseNavbar, can I dry it somehow?
    const itemProps = node.props as DropdownItemProps;
    const itemId = itemProps.itemId ?? itemProps.text;
    return h("a",
        { ...itemProps,
          class: classnames({
            "navbar-item": true,
            "is-active": actualActiveItem.value === itemId,
          }) + [ ` ${node.props?.class}`], // TODO: repeated here and in BaseNavbar, extract it to an util function? Also, it's adding undefined as a class
          onClick: () => actualActiveItem.value = itemId,
        },
        { default: () => itemProps.text }
      );
  } else if (node.type === DropdownDivider) {
    return <hr class="navbar-divider"/>
  } else {
    throw new Error("Children of a NavbarDropdown must be DropdownItems or DropdownDividers!")
  }
}

const slot = useSlots();
const render = () => 
<div class="navbar-dropdown">
  { slot.default && slot.default().map(renderNode) }
</div>

</script>
