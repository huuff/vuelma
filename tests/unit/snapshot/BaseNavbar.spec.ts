import { mount } from "@vue/test-utils";
import BaseNavbar from "@/components/navbar/BaseNavbar.vue";
import NavbarItem from "@/components/navbar/NavbarItem.vue";
import NavbarDropdown from "@/components/navbar/NavbarDropdown.vue";
import DropdownItem from "@/components/dropdown/DropdownItem.vue";
import DropdownDivider from "@/components/dropdown/DropdownDivider.vue";

test("BaseNavbar.vue snapshot", () => {
  const wrapper = mount(BaseNavbar, {
    props: {
      activeItem: "Item 1",
      color: "warning",
    },
    global: {
      components: { NavbarItem, NavbarDropdown, DropdownItem, DropdownDivider },
    },
    slots: {
      brand: `
        <navbar-item titleText="Brand"/>
      `,
      end: `
        <navbar-item titleText="Item 1" />
        <navbar-dropdown>
          <dropdown-item titleText="Item 2" />
          <dropdown-divider />
          <dropdown-item titleText="Item 3" />
        </navbar-dropdown>
      `,
    },
  });

  expect(wrapper.element).toMatchSnapshot();
});
