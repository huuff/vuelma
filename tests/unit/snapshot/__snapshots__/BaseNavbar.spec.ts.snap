// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseNavbar.vue snapshot 1`] = `
<nav
  aria-label="navigation"
  class="navbar is-warning"
  role="navigation"
>
  <div
    class="navbar-brand"
  >
    
    <a
      class="navbar-item"
      titletext="Brand"
    >
      Brand
    </a>
    
    <a
      aria-expanded="false"
      aria-label="menu"
      class="navbar-burger"
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    class="navbar-menu"
  >
    <div
      class="navbar-start"
    >
      <!---->
    </div>
    <div
      class="navbar-end"
    >
      
      <a
        class="navbar-item is-active"
        titletext="Item 1"
      >
        Item 1
      </a>
      <div
        class="navbar-item has-dropdown"
      >
        <a
          class="navbar-link"
        />
        <div
          class="is-right navbar-dropdown"
        >
          <a
            class="navbar-item"
            titletext="Item 2"
          >
            Item 2
          </a>
          <hr
            class="navbar-divider"
          />
          <a
            class="navbar-item"
            titletext="Item 3"
          >
            Item 3
          </a>
        </div>
      </div>
      
    </div>
  </div>
</nav>
`;
